cpp/oneapi/dal/backend/primitives/sort/sort_dpc.cpp

```diff
@@ -0,0 +1,540 @@
+        // valuesOut/indices_bufus);
+    return sycl::event();
+}
+
+template <typename IntType>
+inline void swap(IntType** input, IntType** output) {
+    IntType* tmp = *input;
+    *input = *output;
+    *output = tmp;
+}
+
+template <typename IntType>
+sycl::event radix_sort(sycl::queue& queue,
+                       ndview<IntType, 2>& val_in,
+                       ndview<IntType, 2>& val_out,
+                       ndview<IntType, 2>& buffer,
+                       std::uint32_t sorted_elem_count,
+                       const event_vector& deps) {
+    const std::uint32_t preferable_wg_size = 32;
+    const std::uint32_t expected_buffer_size_for_one_vector = 256;
+
+    ONEDAL_ASSERT(val_in.get_dimension(0) == val_out.get_dimension(0) == buffer.get_dimension(0));
+    ONEDAL_ASSERT(val_in.get_dimension(1) == val_out.get_dimension(1));
+    ONEDAL_ASSERT(buffer.get_dimension(1) == expected_buffer_size_for_one_vector);
+    ONEDAL_ASSERT(sorted_elem_count > 0);
+
+    IntType* labels = val_in.get_mutable_data();
+    IntType* sorted = val_out.get_mutable_data();
+    IntType* radixbuf = buffer.get_mutable_data();
+
+    const std::uint32_t vector_count = dd::integral_cast<std::uint32_t>(val_in.get_dimension(0));
+    const std::uint32_t vector_offset = dd::integral_cast<std::uint32_t>(val_in.get_dimension(1));
+
+    sycl::range<2> global(vector_count, preferable_wg_size);
+    sycl::range<2> local(1, preferable_wg_size);
+
+    sycl::nd_range<2> nd_range2d(global, local);
+
+    auto event = queue.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(deps);
+        cgh.parallel_for(nd_range2d, [=](sycl::nd_item<2> item) {
+            auto sbg = item.get_sub_group();
+            // Code is written for a single subgroup. It's necessary to adjust the local range if idle subgoups are presented
+            if (sbg.get_group_id()[0] > 0) {
+                return;
+            }
+
+            const std::uint32_t global_id = item.get_global_id()[0];
+            const std::uint32_t local_id = item.get_local_id()[1];
+
+            const std::uint32_t local_size = sbg.get_local_range()[0];
+            const std::uint32_t group_aligned_size =
+                sorted_elem_count - sorted_elem_count % local_size;
+            const std::uint32_t rem = sorted_elem_count - group_aligned_size;
+            // radixBuf should be big enough to accumulate radix_range elements
+            const std::uint32_t radix_range = expected_buffer_size_for_one_vector;
+
+            const std::uint32_t radix_count = sizeof(IntType);
+            IntType* input = &labels[global_id * vector_offset];
+            IntType* output = &sorted[global_id * vector_offset];
+            IntType* counters = &radixbuf[global_id * radix_range];
+            //  Radix sort
+            for (std::uint32_t i = 0; i < radix_count; i++) {
+                std::uint8_t* cinput = static_cast<std::uint8_t*>(static_cast<void*>(input));
```

Comment:
Here we can use `reinterpret_cast` as well.

Reply:
fixed